lib/srv/desktop/deskproto/proto.go

```diff
@@ -0,0 +1,254 @@
+}
+
+// MouseMove is the mouse movement message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#3---mouse-move
+type MouseMove struct {
+	X, Y uint32
+}
+
+func (m MouseMove) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeMouseMove))
+	if err := binary.Write(buf, binary.BigEndian, m); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeMouseMove(buf []byte) (MouseMove, error) {
+	var m MouseMove
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &m)
+	return m, trace.Wrap(err)
+}
+
+// MouseButtonType identifies a specific button on the mouse.
+type MouseButtonType byte
+
+const (
+	LeftMouseButton   = MouseButtonType(0)
+	MiddleMouseButton = MouseButtonType(1)
+	RightMouseButton  = MouseButtonType(2)
+)
+
+// ButtonState is the press state of a keyboard or mouse button.
+type ButtonState byte
+
+const (
+	ButtonNotPressed = ButtonState(0)
+	ButtonPressed    = ButtonState(1)
+)
+
+// MouseButton is the mouse button press message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#4---mouse-button
+type MouseButton struct {
+	Button MouseButtonType
+	State  ButtonState
+}
+
+func (m MouseButton) Encode() ([]byte, error) {
+	return []byte{byte(TypeMouseButton), byte(m.Button), byte(m.State)}, nil
+}
+
+func decodeMouseButton(buf []byte) (MouseButton, error) {
+	var m MouseButton
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &m)
+	return m, trace.Wrap(err)
+}
+
+// KeyboardButton is the keyboard button press message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#4---keyboard-input
+type KeyboardButton struct {
+	KeyCode uint32
+	State   ButtonState
+}
+
+func (k KeyboardButton) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeKeyboardButton))
+	if err := binary.Write(buf, binary.BigEndian, k); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeKeyboardButton(buf []byte) (KeyboardButton, error) {
+	var k KeyboardButton
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &k)
+	return k, trace.Wrap(err)
+}
+
+// ClientScreenSpec is the client screen specification.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#1---client-screen-spec
+type ClientScreenSpec struct {
+	Width  uint32
+	Height uint32
+}
+
+func (s ClientScreenSpec) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeClientScreenSpec))
+	if err := binary.Write(buf, binary.BigEndian, s); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeClientScreenSpec(buf []byte) (ClientScreenSpec, error) {
+	var s ClientScreenSpec
+	err := binary.Read(bytes.NewReader(buf[1:]), binary.BigEndian, &s)
+	return s, trace.Wrap(err)
+}
+
+// ClientUsername is the client username.
```

Comment:
Might be helpful to explain in which case this message is sent as well.

Reply:
Message meaning and sequencing is explained in the RFD.
I'd rather keep that in one central place and link to it.